_posts/2022-06-02-tf-2-9/tf-2-9.Rmd

```diff
@@ -0,0 +1,127 @@
+```
+
+```{r, echo = FALSE}
+library(tensorflow)
+library(keras)
+```
+
+## Tensorflow + Keras 2.9
+
+The release of the *Deep Learning with R, 2nd Edition* book coincides
+with new releases of Tensorflow and Keras. These releases bring many
+refinements that allow for more idiomatic and concise R code.
+
+The set of Tensor methods for base R generics has greatly expanded. The
+set of R generics that work with Tensorflow Tensors is now quite
+extensive:
+
+```{r}
+methods(class = "tensorflow.tensor")
+```
+
+```{text}
+ [1] -           !           !=          [           [<-        
+ [6] *           /           &           %/%         %%         
+[11] ^           +           <           <=          ==         
+[16] >           >=          |           abs         acos       
+[21] all         any         aperm       Arg         asin       
+[26] atan        cbind       ceiling     Conj        cos        
+[31] cospi       digamma     dim         exp         expm1      
+[36] floor       Im          is.finite   is.infinite is.nan     
+[41] length      lgamma      log         log10       log1p      
+[46] log2        max         mean        min         Mod        
+[51] print       prod        range       rbind       Re         
+[56] rep         round       sign        sin         sinpi      
+[61] sort        sqrt        str         sum         t          
+[66] tan         tanpi      
+see '?methods' for accessing help and source code
+```
+
+This means that often, you can write the same code for Tensorflow
+Tensors as you would for R arrays. For example, consider this small
+function from Chapter 11 of the book:
+
+```{r}
+reweight_distribution <-
+  function(original_distribution, temperature = 0.5) {
+    original_distribution %>%
+      { exp(log(.) / temperature) } %>%
+      { . / sum(.) }
+  }
+```
+
+Note that functions like `reweight_distribution` work with both 1D R
+vectors and 1D Tensorflow Tensors, since `exp`, `log`, `/`, and `sum`
+are all R generics with methods for Tensorflow Tensors.
+
+In the same vain, this Keras release brings with it a refinement to the
+way custom class extensions to Keras are defined. Partially inspired by
+the new [`R7`](https://rconsortium.github.io/OOP-WG/) syntax, there is a
+new family of functions: `new_layer_class`, `new_model_class`,
+`new_metric_class`, and so on. This new interface substantially
+simplifies the amount of R boilerplate required to define custom Keras
+extensions---a pleasant R interface that serves as a facade over the
+mechanics of sub-classing Python classes. This new interface is the yang
+to the yin of `%py_class%`, a way to mime the Python class definition
+syntax in R. Of course, the "raw" api of converting an `R6Class` to
+Python via `r_to_py()` is still available for users that require full
+control.
+
+This release also brings with it a cornucopia of small improvements
+throughout the entire keras R interface: updated `print` and `plot`
```

Comment:
Like Mara said, I'd add parentheses after function calls (probably got it from her, too :-))